File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const [N, M, V] = input[0].split(' ').map(Number);
let graph = [];
for (let i=1; i<=N; i++) graph[i] = [];
for (let i=1; i<=M; i++){
  let [x,y] = input[i].split(' ').map(Number); 
  graph[x].push(y);
  graph[y].push(x);
}

for (let i=1; i <= N; i++) {
  graph[i].sort((a,b) => a - b);
} 

let visitedD = new Array(N+1).fill(false);

const arrDfs = [];
const arrBfs = [];
const dfs = x => {
    visitedD[x] = true; 
    arrDfs.push(x)
    for (y of graph[x]) {
        if (!visitedD[y]) {
            dfs(y);
        }
    }
}

// 큐 정의
class Queue {
  constructor() {
    this.items = {};
    this.headIndex = 0;
    this.tailIndex = 0;
  }
  enqueue(item) {
    this.items[this.tailIndex] = item;
    this.tailIndex++;
  }
  dequeue() {
    const item = this.items[this.headIndex];
    delete this.items[this.headIndex];
    this.headIndex++;
    return item;
  }
  peek() {
    return this.items[this.headIndex];
  }
  getLength() {
    return this.tailIndex - this.headIndex;
  }
}

// BFS 메서드 정의
// 각 노드가 방문된 정보를 표현
let visitedB = Array(9).fill(false);
function bfs(graph, start, visited) {
  const queue = new Queue();
  queue.enqueue(start);
  // 현재 노드를 방문 처리
  visited[start] = true;
  // 큐가 빌 때까지 반복
  while (queue.getLength() != 0) {
    // 큐에서 하나의 원소를 뽑아 출력하기
    const v = queue.dequeue();
    // console.log(v);
    arrBfs.push(v);
    // 아직 방문하지 않은 인접한 원소들을 큐에 삽입
    for (i of graph[v]) { 
        // graph 배열의 v번째 요소, 노드 v와 직접 연결된 노드들의 목록을 포함. 
        // 예를 들어, graph[1] = [2, 3, 4]일 때, 
        // for (i of graph[1])는 i가 순서대로 2, 3, 4 값을 가지게 된다.
        // i는 인덱스가 아닌 요소의 값
      if (!visited[i]) {
        queue.enqueue(i);
        visited[i] = true;
       
      }
    }
  }
}

// 정의된 DFS 함수 호출
dfs(V);
console.log(arrDfs.join(" "))
// 정의된 BFS 함수 호출
bfs(graph, V, visitedB);
console.log(arrBfs.join(" "))

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
# from collections import deque
# import sys
# input = sys.stdin.read

# data = input().rstrip().split('\n')
# # 정점 개수 N, 간선 개수 M, 탐색 시작 정점 번호 V
# n, m, v = map(int, data[0].split())
# data = data[1:]

# # 인덱스 0 무시
# graph = [[] for i in range(n+1)] 
# # print(graph)
# # 좌표 생성
# for i in range(m):
#     a, b = map(int, data[i].split())
#     graph[a].append(b)
#     graph[b].append(a)
# # print(graph)
# # 정점 번호가 작은것부터 방문이므로 오름차순 정렬
# for adj in graph:
#     adj.sort() 


# def dfs(graph, node, visited): 
#     # 재귀
#     for i in graph[node]:
#         if i not in visited:
#             visited.append(i)
#             dfs(graph, i, visited)
#     return visited

# def bfs(graph, node):
#     # 큐
#     visited = [node]
#     q = deque([node])
#     while q:
#         cur = q.popleft()
#         for i in graph[cur]:
#             if i not in visited:
#                 visited.append(i)
#                 q.append(i)
#     return visited


    

# print(' '.join(map(str, dfs(graph, v, [v]))))
# print(' '.join(map(str, bfs(graph, v))))

# 첫째 줄에 정점의 개수 N(1 ≤ N ≤ 1,000), 간선의 개수 M(1 ≤ M ≤ 10,000), 
# 탐색을 시작할 정점의 번호 V
import sys 
from collections import deque
input = sys.stdin.readline

N, M, V = map(int, input().split())
graph = [[] for _ in range(N+1)]
visited = [False] * (N+1)
# print(visited)
# 다음 M개의 줄에는 간선이 연결하는 두 정점의 번호
for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for list in graph: list.sort()
# print(graph)

def dfs(start): # 스택  
    visited = [False] * (N+1)
    stack = [start]
    # print(start) 
    result = []

    while stack:
        node = stack.pop()
        if not visited[node]: 
            visited[node] = True
            # print(node)
            result.append(node)
            for v in reversed(graph[node]):
                if not visited[v]: 
                    # print(v,999)
                    # visited[v] = True
                    stack.append(v)
    return result

def bfs(start): # 큐
    visited = [False] * (N+1)
    q = deque([start])
    visited[start] =True
    result = []

    while q:
        node = q.popleft()
        result.append(node)
        for v in graph[node]:
            if not visited[v]:
                visited[v] = True
                q.append(v)
    return result



# 정점 번호가 작은 것을 먼저 방문 (오름차순 )
# 더 이상 방문할 수 있는 점이 없는 경우 종료

# 첫째 줄에 DFS, 그 다음 줄에는 BFS
print(' '.join(map(str,dfs(V))))
print(' '.join(map(str,bfs(V))))









